modules/basics/src/main/java/com/opengamma/strata/basics/market/ImmutableReferenceData.java

```diff
@@ -44,21 +45,52 @@
   /**
    * The typed reference data values by identifier.
    */
-  @PropertyDefinition(validate = "notNull", builderType = "Map<? extends ReferenceDataId, TypedReferenceData>")
-  private final ImmutableMap<ReferenceDataId, TypedReferenceData> typedValues;
+  @PropertyDefinition(validate = "notNull", builderType = "Map<? extends ReferenceDataId<?>, ?>")
+  private final ImmutableMap<ReferenceDataId<?>, Object> values;
 
   //-------------------------------------------------------------------------
   /**
-   * Obtains an instance from a map of typed reference data.
+   * Obtains an instance from a map of reference data.
    * <p>
-   * Each identifier in the map refers to an instance of {@link TypedReferenceData}.
-   * This allows multiple values to be connected to the same identifier, accessed by type.
-   * 
-   * @param typedValues  the typed reference data values
-   * @return reference data containing the values
+   * Each entry in the map is a single piece of reference data, keyed by the matching identifier.
+   * For example, a {@code HolidayCalendarId} associated with a {@code HolidayCalendar}.
+   * The caller must ensure that the each entry in the map corresponds with the parameterized
+   * type on the identifier.
+   *
+   * @param values  the reference data values
+   * @return the reference data instance
+   * @throws ClassCastException if a value does not match the parameterized type associated with the key
    */
-  public static ImmutableReferenceData of(Map<? extends ReferenceDataId, TypedReferenceData> typedValues) {
-    return new ImmutableReferenceData(typedValues);
+  public static ImmutableReferenceData of(Map<? extends ReferenceDataId<?>, ?> values) {
+    // validation handles case where value does not match key
+    values.forEach((id, value) -> validateEntry(id, value));
+    return new ImmutableReferenceData(values);
+  }
+
+  /**
+   * Obtains an instance from a single reference data entry.
+   * <p>
+   * This returns an instance containing a single entry based on the specified identifier and value.
+   * This is primarily of interest to test cases.
+   *
+   * @param <T>  the type of the reference data
+   * @param id  the identifier
+   * @param value  the reference data values
+   * @return the reference data instance
+   * @throws ClassCastException if the value does not match the parameterized type associated with the key
+   */
+  public static <T> ImmutableReferenceData of(ReferenceDataId<T> id, T value) {
+    // validation handles edge case where input by raw or polluted types
+    validateEntry(id, value);
+    return new ImmutableReferenceData(ImmutableMap.of(id, value));
+  }
+
+  // validates a single entry
+  private static <T> void validateEntry(ReferenceDataId<?> id, Object value) {
+    if (!id.getReferenceDataType().isInstance(value)) {
```

Comment:
Would be true if `value` is `null`, where throwing a `ClassCastException` would seem a bit misleading. Add specific handling or Javadoc covering null values?